Forbid passing a file base in child apps that are not alone

GiudGiud opened the pull request:
if you did that, it would write to the same file which was very cryptic when it crashed

GiudGiud committed:
Add an error message for child apps outputting to the same files
closes #23743

GiudGiud committed:
Add test for avoiding multiple child app writing to the same file

GiudGiud commented:
Failures unrelated

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1846898/) on c16fccd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25874/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1846916/) on c16fccd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d2dcda</th>
      <th colspan=4>#25874 c16fcc</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.64%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92364</td>
      <td>92372</td>
      <td>+8</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15498</td>
      <td>15493</td>
      <td>-5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25874/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25874/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25874/coverage/xfem)

</details>



This comment will be updated on new commits.

joshuahansel commented on framework/src/outputs/FileOutput.C:
```diff
@@ -64,7 +64,14 @@ FileOutput::FileOutput(const InputParameters & parameters)
+  { // Check that we are the only process or not a subapp
+    if (!_app.isUltimateMaster())
+      if (_app.multiAppNumber() > 0)
+        mooseError("A file base has been specified but there are multiple child apps for this "
+                   "MultiApp. The file base would be obeyed and all child apps would write to the "
+                   "same file. This is disallowed.");
```
```suggestion
        mooseError("The parameter 'file_base' may not be specified for a child app when the MultiApp has multiple instances of the child app.");
```

joshuahansel replied:
Give the reason why as a comment - doesn't need to be in error message.

joshuahansel replied:
Note this suggestion will require clang-formatting.

GiudGiud replied:
why would we not tell the user why we disallow it

joshuahansel replied:
IMO: it's not relevant to the user. All they need to know is: they can't do that. The technical reasons for why are for the developers. However, I don't feel strongly either way, so choose whichever you prefer.

joshuahansel replied:
In general, I just try to not give users more information than they need, to avoid confusion/intimidation.

GiudGiud replied:
I view it differently. If we re going to crash someone s simulation with an error message they need to know as much as possible. If they feel they have to dive into the code to understand we re only dev friendly not user friendly

GiudGiud replied:
now what I have now might not be optimal still

joshuahansel replied:
Updated suggestion

joshuahansel commented on test/tests/outputs/exodus/tests:
```diff
@@ -111,6 +112,29 @@
+  []
+
+  [errors]
+    requirement = "The system shall error if "
```
```suggestion
    requirement = "The system shall report an error if "
```

joshuahansel commented on test/tests/outputs/exodus/tests:
```diff
@@ -111,6 +112,29 @@
+      input = 'exodus_elem_id.i'
+      cli_args = "MultiApps/ma/type=FullSolveMultiApp MultiApps/ma/input_files=exodus_elem_id.i "
+                 "MultiApps/ma/positions='0 0 0 1 1 0' MultiApps/ma/cli_args=Outputs/file_base=bad"
+      expect_err = 'A file base has been specified but there are multiple child apps for this MultiApp'
```
Update this accordingly

joshuahansel commented on test/tests/outputs/exodus/tests:
```diff
@@ -111,6 +112,29 @@
+                 "MultiApps/ma/positions='0 0 0 1 1 0' MultiApps/ma/cli_args=Outputs/file_base=bad"
+      expect_err = 'A file base has been specified but there are multiple child apps for this MultiApp'
+
+      detail = "a file base is passed for file outputs and the child apps would be writing the same "
+               "file."
```
```suggestion
      detail = "a file base is supplied for a child app when there are multiple instances."
```

GiudGiud committed:
Address Josh'review

better requirement & detail
better error message
Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

Update framework/src/outputs/FileOutput.C

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud commented on test/tests/outputs/exodus/tests:
```diff
@@ -111,6 +112,28 @@
+      input = 'exodus_elem_id.i'
+      cli_args = "MultiApps/ma/type=FullSolveMultiApp MultiApps/ma/input_files=exodus_elem_id.i "
+                 "MultiApps/ma/positions='0 0 0 1 1 0' MultiApps/ma/cli_args=Outputs/file_base=bad"
+      expect_err = 'A file base has been specified but there are multiple child apps for this MultiApp'
```
```suggestion
      expect_err = 'The parameter \'file_base\' may not be specified for a child app when the MultiApp has multiple instances of the child app'
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1847401/) on 2354e2c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25874/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25874/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a135b2ecc005d5e4f6319b19c509ca67741f83f7`

joshuahansel commented on framework/src/outputs/FileOutput.C:
```diff
@@ -64,7 +64,14 @@ FileOutput::FileOutput(const InputParameters & parameters)
     _file_num = 0;
 
   if (isParamValid("file_base"))
+  { // Check that we are the only process or not a subapp
```
```suggestion
  {
    // Check that we are the only process or not a subapp
```

joshuahansel replied:
Sorry, just noticed this weird formatting

GiudGiud replied:
fixed

GiudGiud commented:
Thanks for the review!
